src/io/sqlite_store.rs

```diff
@@ -0,0 +1,193 @@
+		let sql =
+			format!("SELECT value FROM {} WHERE namespace=:namespace AND key=:key;", KV_TABLE_NAME);
+
+		let res = locked_conn
+			.query_row(
+				&sql,
+				named_params! {
+					":namespace": namespace,
+					":key": key,
+				},
+				|row| row.get(0),
+			)
+			.map_err(|e| match e {
+				rusqlite::Error::QueryReturnedNoRows => {
+					let msg =
+						format!("Failed to read as key could not be found: {}/{}", namespace, key);
+					std::io::Error::new(std::io::ErrorKind::NotFound, msg)
+				}
+				e => {
+					let msg = format!("Failed to read from key {}/{}: {}", namespace, key, e);
+					std::io::Error::new(std::io::ErrorKind::Other, msg)
+				}
+			})?;
+		Ok(Cursor::new(res))
+	}
+
+	fn write(&self, namespace: &str, key: &str, buf: &[u8]) -> std::io::Result<()> {
+		let locked_conn = self.connection.lock().unwrap();
+
+		let sql = format!(
+			"INSERT OR REPLACE INTO {} (namespace, key, value) VALUES (:namespace, :key, :value);",
+			KV_TABLE_NAME
+		);
+
+		locked_conn
+			.execute(
+				&sql,
+				named_params! {
+					":namespace": namespace,
+					":key": key,
+					":value": buf,
+				},
+			)
+			.map(|_| ())
+			.map_err(|e| {
+				let msg = format!("Failed to write to key {}/{}: {}", namespace, key, e);
+				std::io::Error::new(std::io::ErrorKind::Other, msg)
+			})
+	}
+
+	fn remove(&self, namespace: &str, key: &str) -> std::io::Result<bool> {
+		let sql = format!("DELETE FROM {} WHERE namespace=:namespace AND key=:key;", KV_TABLE_NAME);
+		let changes = self
+			.connection
+			.lock()
+			.unwrap()
+			.execute(
+				&sql,
+				named_params! {
+					":namespace": namespace,
+					":key": key,
+				},
+			)
+			.map_err(|e| {
+				let msg = format!("Failed to delete key {}/{}: {}", namespace, key, e);
+				std::io::Error::new(std::io::ErrorKind::Other, msg)
+			})?;
+
+		let was_present = changes != 0;
+
+		Ok(was_present)
+	}
+
+	fn list(&self, namespace: &str) -> std::io::Result<Vec<String>> {
+		let locked_conn = self.connection.lock().unwrap();
+
+		let sql = format!("SELECT key FROM {} WHERE namespace=:namespace", KV_TABLE_NAME);
+		let mut stmt = locked_conn.prepare(&sql).map_err(|e| {
+			let msg = format!("Failed to prepare statement: {}", e);
+			std::io::Error::new(std::io::ErrorKind::Other, msg)
+		})?;
+
+		let mut keys = Vec::new();
+
+		let rows_iter = stmt
+			.query_map(named_params! {":namespace": namespace, }, |row| row.get(0))
+			.map_err(|e| {
+				let msg = format!("Failed to retrieve queried rows: {}", e);
+				std::io::Error::new(std::io::ErrorKind::Other, msg)
+			})?;
+
+		for k in rows_iter {
+			keys.push(k.map_err(|e| {
+				let msg = format!("Failed to retrieve queried rows: {}", e);
+				std::io::Error::new(std::io::ErrorKind::Other, msg)
+			})?);
+		}
+
+		Ok(keys)
+	}
+}
+
+impl KVStorePersister for SqliteStore {
+	fn persist<W: Writeable>(&self, prefixed_key: &str, object: &W) -> lightning::io::Result<()> {
+		let (namespace, key) = get_namespace_and_key_from_prefixed(prefixed_key)?;
```

Comment:
We shouldn't split the key into a file-path for a non-filesystem-directory datastore. We shouldn't be thinking of this having any relation to file-path or filesystem-directory/subdirectory. This is a flawed assumption reminiscent of filesystem-datastore.

as being discussed in ([link](https://github.com/lightningdevkit/ldk-node/pull/100#discussion_r1207362176))